pkg/sqlmigrations/migrations.go

```diff
@@ -765,6 +772,80 @@ func (m *Manager) StartSchemaChangeJobMigration(ctx context.Context) error {
 	})
 }
 
+// StartSystemNamespaceMigration starts an async task to run the migration that
+// migrates entries from system.namespace (descriptor 2) to system.namespace2
+// (descriptor 30). The task first waits until the upgrade to 20.1 is finalized
+// before running the migration. The migration is retried until it succeeds (on
+// any node).
+func (m *Manager) StartSystemNamespaceMigration(ctx context.Context) error {
+	return m.stopper.RunAsyncTask(ctx, "run-system-namespace-migration", func(ctx context.Context) {
+		log.Info(ctx, "starting wait for upgrade finalization before system.namespace migration")
+		// First wait for the cluster to finalize the upgrade to 20.1. These values
+		// were chosen to be similar to the retry loop for finalizing the cluster
+		// upgrade.
+		waitRetryOpts := retry.Options{
+			InitialBackoff: 10 * time.Second,
+			MaxBackoff:     10 * time.Second,
+			Closer:         m.stopper.ShouldQuiesce(),
+		}
+		for retry := retry.StartWithCtx(ctx, waitRetryOpts); retry.Next(); {
+			if m.settings.Version.IsActive(ctx, clusterversion.VersionNamespaceTableWithSchemas) {
+				break
+			}
+		}
+		select {
+		case <-m.stopper.ShouldQuiesce():
+			return
+		default:
+		}
+		log.VEventf(ctx, 2, "detected upgrade finalization for system.namespace migration")
+
+		// TODO (jordan): add testing knobs here if necessary.
+		if true { //!m.testingKnobs.AlwaysRunNamespaceMigration {
+			// Check whether this migration has already been completed.
+			if kv, err := m.db.Get(ctx, systemNamespaceMigrationKey); err != nil {
+				log.Infof(ctx, "error getting record of schema change job migration: %s", err.Error())
+			} else if kv.Exists() {
+				log.Infof(ctx, "schema change job migration already complete")
+				return
+			}
+		}
+
+		// Now run the migration. This is retried indefinitely until it finishes.
+		log.Infof(ctx, "starting system.namespace migration")
+		r := runner{
+			db:          m.db,
+			codec:       m.codec,
+			sqlExecutor: m.sqlExecutor,
+			settings:    m.settings,
+		}
+		migrationRetryOpts := retry.Options{
+			InitialBackoff: 1 * time.Minute,
+			MaxBackoff:     10 * time.Minute,
+			Closer:         m.stopper.ShouldQuiesce(),
+		}
+		startTime := timeutil.Now().String()
+		for migRetry := retry.Start(migrationRetryOpts); migRetry.Next(); {
+			migrateCtx, _ := m.stopper.WithCancelOnQuiesce(context.Background())
+			migrateCtx = logtags.AddTag(migrateCtx, "system-namespace-migration", nil)
+			if err := migrateSystemNamespace(migrateCtx, r); err != nil {
+				log.Errorf(ctx, "error attempting running system.namespace migration, will retry: %s %s", err.Error(),
+					startTime)
+				continue
+			}
+			log.Infof(ctx, "system.namespace migration completed")
+			if err := m.db.Put(ctx, systemNamespaceMigrationKey, startTime); err != nil {
```

Comment:
So the argument we don't need to write this key as part of the migration txn is that the migration is idempotent, so it doesn't matter how many times it runs?

Reply:
The problem is that different nodes can become 20.1 at different times, I think... I'm not actually sure if there's more problems here. I copied this strategy from the async jobs migration, which is right above this one in the file.